pkg/asset/cluster/openstack/openstack.go

```diff
@@ -35,6 +45,51 @@ func PreTerraform() error {
 	} else if !errors.Is(err, os.ErrNotExist) {
 		return fmt.Errorf("unable to determine if clouds.yaml exists: %w", err)
 	}
+	return nil
+}
+
+// PreTerraform performs any infrastructure initialization which must
+// happen before Terraform creates the remaining infrastructure.
+func PreTerraform(ctx context.Context, tfvarsFile *asset.File, installConfig *installconfig.InstallConfig, clusterID *installconfig.ClusterID, rhcosImage *rhcos_asset.Image) error {
+	if err := replaceBootstrapIgnitionInTFVars(ctx, tfvarsFile, installConfig, clusterID); err != nil {
+		return err
+	}
+
+	if err := tagVIPPorts(ctx, installConfig, clusterID); err != nil {
+		return err
+	}
+
+	// upload the corresponding image to Glance if rhcosImage contains a
+	// URL. If rhcosImage contains a name, then that points to an existing
+	// Glance image.
+	if imageName, isURL := rhcos.GenerateOpenStackImageName(string(*rhcosImage), clusterID.InfraID); isURL {
+		if err := uploadBaseImage(installConfig.Config.Platform.OpenStack.Cloud, rhcosImage, imageName, clusterID, installConfig.Config.Platform.OpenStack.ClusterOSImageProperties); err != nil {
+			return err
+		}
+	}
+
+	return SetTerraformEnvironment()
+}
+
+// PreCAPI performs any infrastructure initialization which must
+// happen before CAPI creates the remaining infrastructure.
+func PreCAPI(ctx context.Context, manifests *[]client.Object, installConfig *installconfig.InstallConfig, clusterID *installconfig.ClusterID, rhcosImage *rhcos_asset.Image) error {
+	if err := replaceBootstrapIgnitionInManifests(ctx, manifests, installConfig, clusterID); err != nil {
+		return err
+	}
+
+	if err := tagVIPPorts(ctx, installConfig, clusterID); err != nil {
+		return err
+	}
+
+	// upload the corresponding image to Glance if rhcosImage contains a
+	// URL. If rhcosImage contains a name, then that points to an existing
+	// Glance image.
+	if imageName, isURL := rhcos.GenerateOpenStackImageName(string(*rhcosImage), clusterID.InfraID); isURL {
+		if err := uploadBaseImage(installConfig.Config.Platform.OpenStack.Cloud, rhcosImage, imageName, clusterID, installConfig.Config.Platform.OpenStack.ClusterOSImageProperties); err != nil {
+			return err
+		}
+	}
```

Comment:
@pierreprinetti Am I correct in thinking this is where we'd slot in things like network creation that we [currently do with TF](https://github.com/openshift/installer/blob/master/data/data/openstack/masters/private-network.tf)?

Reply:
Yes, exactly!